### PR TITLE
[CONF-RUNTIME-EVIDENCE-RUST-DISCOVERY-01] Add Rust runtime evidence discovery

### LINKED ISSUE #1778: [CONF-RUNTIME-EVIDENCE-RUST-DISCOVERY-01] Add Rust toolchain-backed runtime evidence discovery
contract_version: rubin-linear-issue-contract-v1<br>status: OPEN<br>issue_type: executable_conformance_tooling_hardening<br>target_repo: rubin-protocol<br>priority: P2<br>project: Bug/security findings<br>track: conformance<br>parent_issue: 2tbmz9y2xt-lang/rubin-protocol#1772<br>blocked_by: 2tbmz9y2xt-lang/rubin-protocol#1777 Go/common discovery slice must merge first<br>source_refs:

* 2tbmz9y2xt-lang/rubin-protocol#1772 original combined toolchain discovery contract
* 2tbmz9y2xt-lang/rubin-protocol#1777 Go-only/common Part 1 split
* PR #1774 Copilot thread: Rust [mod.rs](<http://mod.rs>) sources cannot be uniquely file-bound by module-prefix filtering
* Prior local reviewer finding: Rust leaf names must be module-bound, not package-global
* Prior Neon pattern: Rust ignored tests must be subtracted from cargo --list evidence

architect_tz_quality_gate:<br>  status: PASS<br>  reason: This child isolates Rust discovery semantics from the Go/common PR so Rust [lib.rs/mod.rs/module/ignored-test](<http://lib.rs/mod.rs/module/ignored-test>) edge cases can be handled without overloading the Part 1 LOC budget.

quality_hardening_addendum:<br>RUB Review Pattern Hardening Matrix applies before implementation and again before commit, push, review-thread reply/resolve, ready, or merge-ready claim against the final diff and claim inventory.

one_invariant:<br>Rust runtime_evidence declared test names must be verified from Cargo/rust test discovery in a source-scope-safe way, and unsupported Rust source paths must fail closed instead of passing through ambiguous module or leaf-name matching.

architecture_class: C - bounded Python tooling extension with Rust/Cargo subprocess boundary<br>system_boundary:<br>This child extends the Part 1 opt-in runtime_evidence discovery plumbing with Rust-only discovery. It must not change Go behavior except where required to use the shared opt-in interface created by 2tbmz9y2xt-lang/rubin-protocol#1777.

layer: conformance tooling / non-consensus validation

max_loc_gate:<br>  max_total_changed_loc: 300<br>  max_production_or_tooling_loc: 190<br>  enforcement: If final diff exceeds either limit, split again before push.

scope:

* tools/check_conformance_edge_pack.py for Rust Cargo discovery backend only
* tools/tests/test_check_conformance_edge_pack.py for Rust discovery contract tests only
* PR body for Rust Part 2 semantics if this opens a new PR
* no conformance baseline, fixture vector, runtime, consensus, README, Go implementation, or unrelated Go discovery changes

accepted_cases:

* Rust declared test names are discovered with cargo test -- --list or an equivalently toolchain-backed Cargo test discovery command.
* Ignored Rust tests are not accepted as active runtime evidence unless the issue explicitly permits ignored evidence.
* Cargo stdout is parsed as full test names; stderr/wrapper noise is not evidence.
* Leaf test names are accepted only when the tool can bind them to the declared Rust source scope without admitting sibling modules or other files.
* Rust [lib.rs](<http://lib.rs>) and [mod.rs](<http://mod.rs>) paths fail closed unless a stronger file-binding strategy is implemented and tested.
* Rust integration-test or non-crate-src paths fail closed unless explicitly supported and tested.
* Missing Cargo/Rust toolchain, timeout, non-zero status, malformed stdout, ambiguous crate resolution, and ambiguous test names fail closed with deterministic errors.

rejected_cases:

* hand-rolled Rust source lexer/parser or regex scanning as proof of executable tests
* accepting a Rust leaf test name from any package/module without source-scope binding
* accepting [mod.rs](<http://mod.rs>) parent-module filtering that can include tests from other files
* accepting [lib.rs](<http://lib.rs>) as file-bound evidence without stronger attribution
* accepting ignored tests as active runtime evidence
* changing consensus, runtime behavior, fixtures, baseline declarations, or RUBIN_L1_CANONICAL
* broad refactor of Go/common discovery from 2tbmz9y2xt-lang/rubin-protocol#1777
* exceeding the LOC gate and pushing anyway

hostile_cases:

* Rust module contains tests in sibling/submodule files with the same leaf test name
* [mod.rs](<http://mod.rs>) contains or parents nested modules that produce matching cargo test names from other files
* [lib.rs](<http://lib.rs>) exposes module-qualified tests that are not file-bound to [lib.rs](<http://lib.rs>)
* ignored tests appear in normal cargo list output unless subtracted
* cargo emits warnings, build script output, or harness text
* cargo discovery succeeds but ignored-test discovery fails
* Cargo.toml [package.name](<http://package.name>) is missing, malformed, or unreadable

planned_execution_config:<br>primary_profile: python-tooling<br>secondary_profiles: rust-test-discovery only through Cargo/Rust toolchain commands<br>branch: new child branch after 2tbmz9y2xt-lang/rubin-protocol#1777 merges<br>pr: new PR after 2tbmz9y2xt-lang/rubin-protocol#1777; do not extend PR #1774 unless controller explicitly asks to re-merge scopes<br>implementation_steps:

* start from origin/main after 2tbmz9y2xt-lang/rubin-protocol#1777 merges
* reuse Part 1 opt-in discovery plumbing without broad rewrite
* implement Rust package/crate resolution and Cargo command boundary
* add Rust command-output tests for module binding, ignored subtraction, [lib.rs](<http://lib.rs>) fail-closed, [mod.rs](<http://mod.rs>) fail-closed, and non-src fail-closed
* run targeted Rust discovery smoke commands only for current runtime_reorg evidence paths

mandatory_local_gates:

* python3 -m unittest tools.tests.test_check_conformance_edge_pack
* python3 -m py_compile tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py
* ruff check tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py
* bandit -q tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py
* vulture tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py --min-confidence 90
* scripts/dev-env.sh -- python3 tools/check_conformance_edge_pack.py
* scripts/dev-env.sh -- python3 tools/gen_conformance_matrix.py --check
* scripts/dev-env.sh -- python3 tools/check_conformance_fixtures_policy.py
* targeted Rust test discovery command documented by the PR
* git diff --check origin/main

pre_push_review:<br>After candidate commit only: run /Users/gpt/bin/rubin-precommit-one-review --include-base-diff --light-python, then exactly one fresh stock code-review subagent using the wrapper pointer, fork_context=false, then close the subagent. No manual pre-push-reviewer, no fanout, no PR-thread context.

done_when:

* Rust runtime_evidence discovery is Cargo/toolchain-backed or explicitly fails closed for unsupported Rust source scopes.
* Rust [lib.rs](<http://lib.rs>) and [mod.rs](<http://mod.rs>) false-positive paths cannot satisfy file-bound evidence unless a stronger binding strategy is implemented and tested.
* Ignored Rust tests are not accepted as active runtime evidence.
* Mandatory local gates pass.
* Final diff satisfies max_loc_gate.
* PR body states Rust Part 2 semantics and non-goals.

not_done_until:

* no unresolved review thread remains on current PR head after fresh GraphQL poll
* GitHub checks are green or explicitly classified with evidence
* any new reusable failure class from this child is recorded in Neon as type=pattern

Source of truth: this Linear child issue. 2tbmz9y2xt-lang/rubin-protocol#1772 is the split parent; 2tbmz9y2xt-lang/rubin-protocol#1777 is the required predecessor.

### PR DESCRIPTION
**Pull request by 2tbmz9y2xt-lang**
Invariant:
Rust `runtime_evidence` declared test names can be verified with opt-in Cargo-backed discovery only when the source scope is unambiguous and the test is active, not ignored.

Layer:
Non-consensus conformance tooling.

Files changed:
- `tools/check_conformance_edge_pack.py`
- `tools/tests/test_check_conformance_edge_pack.py`

Tests:
- `python3 -m unittest tools.tests.test_check_conformance_edge_pack`
- `python3 -m py_compile tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py`
- `ruff check tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py`
- `bandit -q tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py`
- `vulture tools/check_conformance_edge_pack.py tools/tests/test_check_conformance_edge_pack.py --min-confidence 90`
- `scripts/dev-env.sh -- python3 tools/check_conformance_edge_pack.py`
- `scripts/dev-env.sh -- python3 tools/check_conformance_edge_pack.py --verify-runtime-evidence-rust`
- `scripts/dev-env.sh -- python3 tools/check_conformance_edge_pack.py --verify-runtime-evidence-go --verify-runtime-evidence-rust`
- `scripts/dev-env.sh -- python3 tools/gen_conformance_matrix.py --check`
- `scripts/dev-env.sh -- python3 tools/check_conformance_fixtures_policy.py`
- `git diff --check origin/main`
- `rubin-precommit-one-review --base-ref origin/main --include-base-diff --light-python`
- stock code-review subagent on `origin/main...HEAD`: `No findings`
- `rubin-push --no-coverage-reason "RUB-350 changes only Python conformance tooling/tests for Rust runtime evidence discovery; no Go/Rust coverable client source changed"`

Behavior impact:
- Default `tools/check_conformance_edge_pack.py` remains schema/path/accounting-only and does not run Go or Cargo.
- New opt-in `--verify-runtime-evidence-rust` verifies Rust entries with `cargo test --locked -p <package> -- --list`, subtracts `--ignored --list`, and binds accepted leaf names to a non-ambiguous `src/<module>.rs` module prefix.
- Cargo discovery runs with a temp `CARGO_TARGET_DIR`, `CARGO_INCREMENTAL=0`, and ambient Rust/Cargo wrapper/flag env unset.

Compatibility impact:
No consensus, runtime, fixture, baseline, Go implementation, or Rust implementation behavior changes.

Known non-goals:
- No Rust `lib.rs`, `main.rs`, `mod.rs`, `src/bin`, or integration-test source-scope support in this PR; those fail closed.
- No hand-rolled Rust source parser.
- No generated artifact updates.

Risk:
Opt-in Rust discovery compiles the package test harness for listing. It is bounded by the existing discovery timeout and uses `--locked` plus an isolated target dir.

Rollback:
Revert this PR to remove the Rust discovery flag and backend while preserving the existing Go discovery from the prior slice.

Not checked:
Full Rust workspace tests were not run because this PR changes only Python conformance tooling/tests and does not change Rust client code.

Closes #1778

### CODE REVIEW

File: tools/check_conformance_edge_pack.py [hunk 1]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+    if not source_parts or source_parts[0] == "bin" or source_parts[-1] in {"lib.rs", "main.rs", "mod.rs"}:
+        return Path(), "", "unsupported Rust runtime evidence source scope"
+    module_parts = list(source_parts)
+    module_parts[-1] = Path(module_parts[-1]).stem
+    return crate_root, "::".join(module_parts), None
+
+
+def rust_listed_tests(output: str, module_scope: str) -> set[str]:
+    tests: set[str] = set()
+    prefix = f"{module_scope}::"
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 (and the earlier 72434a1 thread-fix path). Rust discovery no longer accepts a broad module prefix: it uses Cargo --lib discovery, subtracts ignored full paths, requires direct module_scope::tests::<leaf>, and cross-checks the leaf against the referenced source file inline tests. Regressions cover wrong module, nested module paths, ignored tests, external mod tests, and unsupported source scopes.

File: tools/check_conformance_edge_pack.py [hunk 2]
```diff
@@ -154,7 +168,8 @@ def run_discovery_command(cmd: list[str], *, cwd: Path, command_runner, env: dic
         return "", f"runtime_evidence discovery command failed: {cmd[0]} exit {completed.returncode}"
     if not isinstance(completed.stdout, str) or not isinstance(completed.stderr, str):
         return "", "runtime_evidence discovery command produced malformed stdout"
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 via the existing thread-fix path. The subprocess-output diagnostic now says stdout/stderr because both streams are validated before output handling.

File: tools/check_conformance_edge_pack.py [hunk 3]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+        return rust_file_discovered_tests(raw_path, repo_root=repo_root, expected_tests=expected_tests, command_runner=command_runner)
     return set(), None
 
 
+def rust_package_name(crate_root: Path) -> tuple[str, str | None]:
+    cargo_toml = crate_root / "Cargo.toml"
+    try:
+        lines = cargo_toml.read_text(encoding="utf-8").splitlines()
+    except (OSError, UnicodeDecodeError) as exc:
+        return "", f"runtime_evidence discovery command produced unreadable Cargo.toml: {exc}"
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74. Cargo.toml filesystem/encoding failures now report runtime_evidence could not read Cargo.toml, TOML syntax failures report could not parse Cargo.toml, and missing/malformed package.name keeps the discovery-name error.

File: tools/check_conformance_edge_pack.py [hunk 4]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+        )
+        if error is not None:
+            return set(), error
+        ignored_output, ignored_error = run_discovery_command(
+            ["cargo", "test", "--locked", "-p", package, "--", "--ignored", "--list"],
+            cwd=rust_root,
+            env=env,
+            command_runner=command_runner,
+            merge_stderr=False,
+        )
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 via the existing thread-fix path. Rust cargo --list results are cached per package for one runtime_evidence pass; the regression covers two files in rubin-node and asserts one normal list and one ignored-list cargo call.

File: tools/check_conformance_edge_pack.py [hunk 5]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+    tests: set[str] = set()
+    prefix = f"{module_scope}::"
+    for line in output.splitlines():
+        stripped = line.strip()
+        match = re.match(r"([^:][^ ]*(?:::[^ ]+)*)\: test$", stripped)
+        if match is None:
+            continue
+        test_path = match.group(1)
+        if test_path.startswith(prefix):
+            tests.add(test_path.rsplit("::", 1)[-1])
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 (same code path as 72434a1). The matcher is now file-attributed: Cargo --lib output must match direct module_scope::tests::<leaf>, and the leaf must also be declared in the referenced source file inline tests module. Nested modules and external mod tests are rejected by regression coverage.

File: tools/check_conformance_edge_pack.py [hunk 6]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+    for line in lines:
+        stripped = line.strip()
+        if stripped == "[package]":
+            in_package = True
+            continue
+        if stripped.startswith("[") and stripped.endswith("]"):
+            in_package = False
+            continue
+        if in_package:
+            match = re.match(r'name\s*=\s*"([^"]+)"\s*$', stripped)
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74. rust_package_name now uses tomllib on Python 3.11+ or tomli on older Python, reads [package].name through TOML semantics, and the regression fixture uses a single-quoted name with an inline comment.

File: tools/check_conformance_edge_pack.py [hunk 7]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+        return set(), scope_error
+    package, package_error = rust_package_name(repo_root / crate_rel)
+    if package_error is not None:
+        return set(), package_error
+    rust_root = repo_root / "clients" / "rust"
+    with tempfile.TemporaryDirectory(prefix="rubin-rust-test-list-") as temp_dir:
+        env = rust_discovery_env(temp_dir)
+        output, error = run_discovery_command(
+            ["cargo", "test", "--locked", "-p", package, "--", "--list"],
+            cwd=rust_root,
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 via the existing thread-fix path. Rust discovery invokes cargo test --locked -p <package> --lib -- --list and the ignored-list equivalent, so integration/non-library test targets cannot satisfy src/... runtime evidence.

File: tools/check_conformance_edge_pack.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 via the existing thread-fix path. Ignored filtering now subtracts fully qualified test paths before direct module_scope::tests::<leaf> matching, so unrelated same-leaf names do not mask active tests.

File: tools/check_conformance_edge_pack.py [hunk 8]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+) -> tuple[set[str], str | None]:
+    crate_rel, module_scope, scope_error = rust_module_scope(raw_path)
+    if scope_error is not None:
+        return set(), scope_error
+    package, package_error = rust_package_name(repo_root / crate_rel)
+    if package_error is not None:
+        return set(), package_error
+    rust_root = repo_root / "clients" / "rust"
+    with tempfile.TemporaryDirectory(prefix="rubin-rust-test-list-") as temp_dir:
+        env = rust_discovery_env(temp_dir)
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 via the existing thread-fix path. Runtime evidence discovery now reuses a per-package cargo cache and shared temp Cargo target dir for the pass; the regression asserts one normal and one ignored cargo list call for two Rust files in the same crate.

File: tools/check_conformance_edge_pack.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in 93cbc74 (same file-attribution path as 72434a1). The current matcher requires direct module_scope::tests::<leaf> plus source-file inline test binding, and regression coverage rejects child modules and external mod tests.

File: tools/check_conformance_edge_pack.py [hunk 9]
```diff
@@ -252,18 +279,113 @@ def discovered_runtime_tests(
...
+            in_package = True
+            continue
+        if stripped.startswith("[") and stripped.endswith("]"):
+            in_package = False
+            continue
+        if in_package:
+            match = re.match(r'name\s*=\s*"([^"]+)"\s*$', stripped)
+            if match:
+                return match.group(1), None
+    return "", "runtime_evidence discovery command could not determine Rust package name"
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py:331]: Fixed in 93cbc74. The regex fallback was removed: Cargo.toml package metadata is parsed with tomllib/tomli only, and the Rust opt-in path fails closed if neither TOML parser is available. The regression fixture covers single-quoted package.name with an inline TOML comment.

File: tools/check_conformance_edge_pack.py [hunk 10]
```diff
@@ -165,6 +191,24 @@ def go_discovery_env(temp_dir: str) -> dict[str, str]:
...
+        "CARGO_BUILD_RUSTFLAGS": None,
+        "CARGO_ENCODED_RUSTFLAGS": None,
+        "CARGO_NET_OFFLINE": None,
+        "RUSTC": None,
+        "RUSTC_WRAPPER": None,
+        "RUSTC_WORKSPACE_WRAPPER": None,
+        "RUSTUP_HOME": os.environ.get("RUSTUP_HOME", str(Path.home() / ".rustup")),
+        "RUSTFLAGS": None,
+    }
+
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py:208]: Fixed in e40551e. rust_discovery_env no longer calls Path.home(); RUSTUP_HOME is taken only from os.environ.get("RUSTUP_HOME") and is left unset when absent. Added a regression that patches Path.home to raise RuntimeError and verifies env construction still succeeds.

File: tools/check_conformance_edge_pack.py [hunk 11]
```diff
@@ -165,6 +183,24 @@ def go_discovery_env(temp_dir: str) -> dict[str, str]:
     }
 
 
+def rust_discovery_env(temp_dir: str) -> dict[str, str | None]:
+    return {
+        "CARGO_INCREMENTAL": "0",
+        "CARGO_HOME": str(Path(temp_dir) / "cargo-home"), "HOME": temp_dir,
+        "CARGO_TARGET_DIR": str(Path(temp_dir) / "target"),
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in e40551e. Rust discovery now keeps inherited CARGO_HOME, HOME, and CARGO_NET_OFFLINE so cargo can use prewarmed caches/custom registry config, while still forcing a temp CARGO_TARGET_DIR and scrubbing build-affecting env.

File: tools/check_conformance_edge_pack.py [hunk 12]
```diff
@@ -252,18 +288,150 @@ def discovered_runtime_tests(
...
+
+
+def rust_source_declared_tests(source_path: Path) -> tuple[set[str], str | None]:
+    try:
+        source = source_path.read_text(encoding="utf-8")
+    except (OSError, UnicodeDecodeError) as exc:
+        return set(), f"runtime_evidence could not read Rust source: {exc}"
+    if re.search(r"(?m)^\s*(?:#\[[^\n]*\]\s*)*mod\s+tests\s*;", source):
+        return set(), "unsupported Rust runtime evidence source scope"
+    match = re.search(r"(?s)#\s*\[\s*cfg\s*\(\s*test\s*\)\s*\]\s*mod\s+tests\s*\{(?P<body>.*)\n\}", source)
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in e40551e. rust_source_declared_tests no longer uses a greedy body regex; it slices the inline tests module with balanced braces before scanning nested modules/tests. Added a regression with mod helper after the tests block.

File: tools/check_conformance_edge_pack.py [hunk 13]
```diff
@@ -252,18 +288,150 @@ def discovered_runtime_tests(
...
+        return set(), scope_error
+    package, package_error = rust_package_name(repo_root / crate_rel)
+    if package_error is not None:
+        return set(), package_error
+    rust_root = repo_root / "clients" / "rust"
+    for root in [rust_root, *rust_root.parents]:
+        if (root / ".cargo" / "config").exists() or (root / ".cargo" / "config.toml").exists():
+            return set(), "unsupported Rust runtime evidence Cargo config"
+        if root == repo_root:
+            break
```
[the remark this answers is not in the document]
Comment (2tbmz9y2xt-lang) [reply tools/check_conformance_edge_pack.py]: Fixed in e40551e. The .cargo/config(.toml) guard now scans clients/rust and all parent directories instead of stopping at repo_root. Added regression coverage for a parent .cargo/config.toml outside repo_root.
